added the function and tests

mikecann opened the pull request:
As discussed in: https://github.com/kentcdodds/nps-utils/issues/22

For this PR, I support the happy-path if you are using the standard yarn workspaces or lerna structure:

```
module.exports = {
  scripts: {
    client: importPackage("client"),
    server: importPackage("server"),
    shared: importPackage("shared"),
  }
}
```

Also support you optionally providing the path:

```
module.exports = {
  scripts: {
    client: importPackage({path: "./mycustomfolder/my-nps-scripts.js" })
  }
}
```

It detects "descriptions" and handles them correctly.

Wasnt too sure how to structure the required test files (client, server). 

Fixed most of the es-lint issues but one, no arrow functions?

Plus i'm not sure how to actually functionally test this other than using console.log, perhaps you can advise?

mikecann commented:
Probably this should respect the `options` object of the imported script too

kentcdodds commented on src/__snapshots__/index.test.js.snap:
```diff
@@ -28,6 +28,14 @@ exports[`ifWindows as darwin 1`] = `"echo alternate"`;
 
 exports[`ifWindows as win32 1`] = `"echo main"`;
 
+exports[`includePackage as darwin 1`] = `"{\\"root\\":\\"cd packages/client && npm start root\\",\\"nested\\":{\\"child1\\":\\"cd packages/client && npm start nested.child1\\",\\"child2\\":{\\"script\\":\\"cd packages/client && npm start nested.child2 && cd /\\"C://dev//me//experiments//nps-utils/\\"\\",\\"description\\":\\"this is a description\\"},\\"child3\\":{\\"deep\\":{\\"script\\":\\"cd pa ...
```
Looks like you're on a windows machine, so that's why the snapshots are failing. We need to serialize it differently to make sure things are working properly regardless of the platform. That's what `relativeizePath` is supposed to do. Could you figure out why it's not doing that?

Actually... Now that I think of it, I think this is a bit of a different kind of feature so I don't think I want to take a snapshot of it like we do with the others. Why don't you add a `test` block at the bottom of the file rather than include it in the `snapshotTests` object...

mikecann replied:
Ye I thought so. It doesn't really seem to fit the snapshots quite so much. Ill write some tests for it instead.

mikecann replied:
Im struggling with testing this as im not sure how to mock "require" correctly.

```import {includePackage} from './index'

test('it runs the default on a package', () => {
  const packageName = 'foo'
  jest.mock(`./packages/${packageName}/package-scripts.js`, () => ({
    scripts: {
      default: 'echo hello',
    },
  }))
  const outp = includePackage(packageName)
  expect(outp).toBe('{"default": "cd packages/foo && npm start default"')
})
```

Any tips?

kentcdodds replied:
Hi @mikecann, don't put a `jest.mock` call inside a `test`. It doesn't work like that. Put it at the top of the file. Also, I'm not entirely sure how things work with mocking files that don't actually exist... You may have to make a `__fixtures__` directory and put real files in there.

I'm sorry. I have very limited time to give to maintaining this package :-/ Good luck.

kentcdodds commented on src/index.js:
```diff
@@ -326,6 +327,67 @@ function crossEnv(args) {
+  try {
+    scripts = require(packageScriptsPath)
+  } catch (e) {
+    throw Error(commonTags.oneLine`Couldnt include the package, 
```
This is helpful to have a good error message, but let's only throw it if we're sure the reason the original error was thrown is because the file does not exist (maybe do a require.resolve first or check the type of error). I've been bitten by this in the past where the error thrown was actually a syntax error in the file and I spent a lot of time trying to figure out why it couldn't find my file when it turned out that it could and the error was a red herring.

mikecann replied:
Oh thats a very good point. Ill write some tests for syntax errors too.

mikecann replied:
Okay unfortunately require.resolve doesnt work with jest mocks, so decided that the extra complication probably isnt worth it and the native errors are good enough.

kentcdodds commented on src/index.js:
```diff
@@ -326,6 +327,67 @@ function crossEnv(args) {
+  
+  const startingDir = process.cwd()
+  const relativeDir = path.relative(startingDir,
+    path.dirname(packageScriptsPath)).split('\\').join('/')
```
Will this work the same on Windows and Linux?

mikecann replied:
Ah you know what, I think thats why I was having travis issues. Ill look at it now.

kentcdodds commented on src/index.js:
```diff
@@ -326,6 +327,67 @@ function crossEnv(args) {
+  }
+ 
+  const outp = replace(scripts.scripts, '')
+  //console.log(JSON.stringify(outp,null,2));
```
Let's remove this 👍

mikecann replied:
oops, yes, sorry

kentcdodds commented on src/index.test.js:
```diff
@@ -55,6 +55,11 @@ const snapshotTests = {
+  //  JSON.stringify(includePackage('client')),
+  // includePackageWithPath: ({includePackage}) =>
+  //   JSON.stringify(includePackage({path:
+  //     './packages/someotherstructure/server/some-other-scripts-name.js'})),
```
Let's remove these

kentcdodds commented on src/index.test.js:
```diff
@@ -68,6 +73,8 @@ Object.keys(snapshotTests).forEach(testName => {
   })
 })
 
+
```
Let's remove these extra lines

mikecann replied:
yep sorry, not a good citizen

kentcdodds commented on src/packages/client/package-scripts.js:
```diff
@@ -0,0 +1,18 @@
+module.exports = {
```
Is this file necessary?

mikecann replied:
nope, removed

kentcdodds commented on src/packages/someotherstructure/server/some-other-scripts-name.js:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
I'd this file necessary?

kentcdodds replied:
If these files are necessary, let's add a comment to them indicating why they exist 👍

mikecann replied:
nope, sorry they arent needed any more

kentcdodds commented on src/includePackage.test.js:
```diff
@@ -0,0 +1,101 @@
+    scripts: {
+      default: 'echo foo',
+    },
+  }), {virtual: true})
```
Oooohhhh! `virtual: true` is something new to me! Cool! So does this mean you can mock files that don't exist? This is very useful. TIL!

mikecann replied:
Yep! However a gotcha is they seem to stay alive for the entire lifetime of all the tests not just the scope of that one test. Hence why I was careful to have a unique module path each time I used it.

kentcdodds commented:
Would you like to [add yourself](../blob/master/CONTRIBUTING.md#add-yourself-as-a-contributor) to the [contributors table](../#contributors)?

mikecann commented:
@kentcdodds thank you so much for the review, really helpful. I have fixed everything that I could. The only thing is those helpful error messages had to go due to limitations in jest mocking, the native errors are okay.

I have updated the readme to add the new method, I hope is is okay. Im not sure how the docs get generated so couldnt check that they look okay.

I ran into issues automatically testing this on multiple platforms, a simple solution was to replace '\\' with '/' which seems to work on windows just fine these days.

kentcdodds commented:
This is great. Thank you so much for iterating and teaching me through this process!

mikecann commented:
@kentcdodds the pleasure was all mine!

kentcdodds commented:
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

mikecann commented:
Thats awesome thanks Kent. Just one thing, do the docs auto-generate after some time?

kentcdodds commented:
Well... It's supposed to, but it never really worked well. It looks like things have been updated on the esdoc site and it seems to be working mostly, but it doesn't have your new method. I'm not sure why not. Why don't we give it a day and see. If it doesn't update then maybe we should just put docs in the README...

mikecann commented:
Sure, sounds like a plan, perhaps they have some cronjob to check for updates.
